spec/mempool/gossip/p2p.qnt

```diff
@@ -0,0 +1,141 @@
+// -*- mode: Bluespec; -*-
+/*******************************************************************************
+This module specifies the P2P layer as needed for the gossip protocols. It
+includes the definitions of nodes, peers, network topology, sending messages,
+nodes joining and leaving the network.
+*******************************************************************************/
+module p2p {
+    import spells.* from "./spells"
+
+    // A node identifier.
+    type NodeID = str
+
+    //--------------------------------------------------------------------------
+    // Parameters
+    //--------------------------------------------------------------------------
+
+    // The set of all possible node IDs, even those that are not initially
+    // connected to the network.
+    const NodeIDs: Set[NodeID]
+
+    // Initial network topology. A topology is defined by the set of peers each
+    // node has.
+    const InitialPeers: NodeID -> Set[NodeID]
+
+    //--------------------------------------------------------------------------
+    // State
+    //--------------------------------------------------------------------------
+    // To model network communication, each node has a queue (a list) of
+    // incoming messages. Node A sends a message to a node B by appending the
+    // message to B's queue. We use queues to model that messages arrive in
+    // order. Messages have a sender (a node ID).
```

Comment:
```suggestion
    // FIFO order. Messages have a sender (a node ID).
```

Reply:
Do we really need this assumption?

Reply:
Yes, we assume that the transport layer guarantees a message ordering. That's why the type of the incoming messages is a List instead of a Set. I'll clarify in the text.